### PR TITLE
feat(api): add project secrets

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-cds#19900.0 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19900.0 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

File: cli/cdsctl/experimental_project_secret.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+	"github.com/spf13/cobra"
+
+	"github.com/ovh/cds/cli"
+	"github.com/ovh/cds/sdk"
+)
+
+var projectSecretCmd = cli.Command{
+	Name:    "secret",
+	Aliases: []string{""},
```
Comment (yesnault) [inline cli/cdsctl/experimental_project_secret.go:14]:
```suggestion
	Aliases: []string{"secrets"},
```

File: cli/cdsctl/experimental_project_secret.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	Short:   "Manage secret on a CDS project",
```
Comment (yesnault) [inline cli/cdsctl/experimental_project_secret.go:15]:
```suggestion
	Short:   "Manage secrets on a CDS project",
```

File: cli/cdsctl/experimental_project_secret.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		cli.NewCommand(projectSecretAddCmd, projectSecretAddFunc, nil, withAllCommandModifiers()...),
+		cli.NewListCommand(projectSecretListCmd, projectSecretListFunc, nil, withAllCommandModifiers()...),
+		cli.NewDeleteCommand(projectSecretDeleteCmd, projectSecretDeleteFunc, nil, withAllCommandModifiers()...),
+		cli.NewCommand(projectSecretUpdateCmd, projectSecretUpdateFunc, nil, withAllCommandModifiers()...),
+	})
+}
+
+var projectSecretListCmd = cli.Command{
+	Name:  "list",
+	Short: "List secret available on a project",
```
Comment (yesnault) [inline cli/cdsctl/experimental_project_secret.go:29]:
```suggestion
	Short: "List secrets available on a project",
```

**Review by yesnault — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-cds#19935.0 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19935.0 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

File: engine/api/v2_project_secret.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+			if err != nil {
+				return err
+			}
+
+			return service.WriteJSON(w, secrets, http.StatusOK)
+		}
+}
+
+func (api *API) deleteProjectSecretHandler() ([]service.RbacChecker, service.Handler) {
+	return service.RBAC(api.projectManage),
```
Comment (fsamin) [inline engine/api/v2_project_secret.go:37]: maybe we should a have a permission for secrets

File: engine/api/v2_project_secret.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+			if err := tx.Commit(); err != nil {
+				return sdk.WithStack(err)
+			}
+			event_v2.PublishProjectSecretEvent(ctx, api.Cache, sdk.EventSecretDeleted, *secret, *u.AuthConsumerUser.AuthentifiedUser)
+
+			return nil
+		}
+}
+func (api *API) putProjectSecretHandler() ([]service.RbacChecker, service.Handler) {
+	return service.RBAC(api.projectManage),
```
Comment (fsamin) [inline engine/api/v2_project_secret.go:77]: same

File: engine/api/v2_project_secret.go [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+			if u == nil {
+				return sdk.WithStack(sdk.ErrForbidden)
+			}
+
+			var secret sdk.ProjectSecret
+			if err := service.UnmarshalBody(req, &secret); err != nil {
+				return err
+			}
+
+			reg, err := regexp.Compile(sdk.SecretNamePattern)
```
Comment (fsamin) [inline engine/api/v2_project_secret.go:93]: use mustCompile and only once

File: engine/api/v2_project_secret.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
```
Comment (fsamin) [inline engine/api/v2_project_secret.go:155]: same

File: engine/api/v2_queue_worker.go [hunk 8]
```diff
@@ -74,7 +80,7 @@ func (api *API) postV2WorkerTakeJobHandler() ([]service.RbacChecker, service.Han
 		if err != nil {
 			return err
 		}
-		secrets := computeSecretsContext(ctx, *projWithSecrets, *vcsWithSecrets)
+		secrets := computeSecretsContext(ctx, *projWithSecrets, projectSecrets, *vcsWithSecrets)
```
Comment (fsamin) [inline engine/api/v2_queue_worker.go:83]: Are all the secrets injected for all workflow runs, without any declaration ?

**Review by fsamin — CHANGES_REQUESTED** [review]

**Pull request closed without merging**